flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/catalog/CatalogCalciteSchema.java

```diff
@@ -0,0 +1,248 @@
+ */
+
+package org.apache.flink.table.catalog;
+
+import org.apache.flink.annotation.Internal;
+import org.apache.flink.table.catalog.exceptions.CatalogException;
+import org.apache.flink.table.catalog.exceptions.DatabaseNotExistException;
+import org.apache.flink.table.catalog.exceptions.TableNotExistException;
+
+import org.apache.calcite.linq4j.tree.Expression;
+import org.apache.calcite.rel.type.RelDataType;
+import org.apache.calcite.rel.type.RelDataTypeFactory;
+import org.apache.calcite.rel.type.RelProtoDataType;
+import org.apache.calcite.schema.Function;
+import org.apache.calcite.schema.Schema;
+import org.apache.calcite.schema.SchemaPlus;
+import org.apache.calcite.schema.SchemaVersion;
+import org.apache.calcite.schema.Schemas;
+import org.apache.calcite.schema.Table;
+import org.apache.calcite.sql.type.SqlTypeName;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.Set;
+
+/**
+ * A mapping between Flink's catalog and Calcite's schema. This enables to look up and access tables
+ * in SQL queries without registering tables in advance. Databases are registered as sub-schemas in the schema.
+ * This mapping is modeled as a strict two-level reference structure for Flink in Calcite,
+ * the full path of tables and views is of format [catalog_name].[db_name].[meta-object_name].
+ */
+@Internal
+public class CatalogCalciteSchema implements Schema {
+	private static final Logger LOGGER = LoggerFactory.getLogger(CatalogCalciteSchema.class);
+
+	private final String catalogName;
+	private final ReadableCatalog catalog;
+
+	public CatalogCalciteSchema(String catalogName, ReadableCatalog catalog) {
+		this.catalogName = catalogName;
+		this.catalog = catalog;
+	}
+
+	/**
+	 * Looks up a sub-schema (database) by the given sub-schema name.
+	 *
+	 * @param schemaName Name of sub-schema to look up.
+	 * @return Sub-schema with a given dbName, or null.
+	 */
+	@Override
+	public Schema getSubSchema(String schemaName) {
+
+		if (catalog.databaseExists(schemaName)) {
+			return new DatabaseCalciteSchema(schemaName, catalog);
+		} else {
+			LOGGER.error(String.format("Schema %s does not exist in catalog %s", schemaName, catalogName));
+			throw new CatalogException(
+				new DatabaseNotExistException(catalogName, schemaName));
+		}
+	}
+
+	@Override
+	public Set<String> getSubSchemaNames() {
+		return new HashSet<>(catalog.listDatabases());
+	}
+
+	@Override
+	public Table getTable(String name) {
+		throw new UnsupportedOperationException(
+			"CatalogCalciteSchema does not support getTable()");
+	}
+
+	@Override
+	public Set<String> getTableNames() {
+		throw new UnsupportedOperationException(
+			"CatalogCalciteSchema does not support getTableNames()");
+	}
+
+	@Override
+	public RelProtoDataType getType(String name) {
+		return new RelProtoDataType() {
+			@Override
+			public RelDataType apply(RelDataTypeFactory relDataTypeFactory) {
```

Comment:
Shouldn't it return only user defined types that this catalog contain?

Reply:
Right now we haven't support user defined types in catalog yet. This is how Blink handles it. I can change it to return `null` to match existing `ExternalCatalogSchema` in flink-table-planner. What do you think?

Reply:
I would prefer returning `null`, otherwise it is a bit misleading for me, because as far as I understand the `Schema` interface it is only for user/catalog specific types.

Reply:
ok, sounds good